src/os/bluestore/BlueStore.cc

```diff
@@ -12281,22 +12283,22 @@ void BlueStore::_zoned_cleaner_thread() {
   ceph_assert(!zoned_cleaner_started);
   zoned_cleaner_started = true;
   zoned_cleaner_cond.notify_all();
-  std::deque<uint64_t> zones_to_clean;
+  const auto *zones_to_clean = shared_alloc.a->zoned_get_zones_to_clean();
   while (true) {
-    if (zoned_cleaner_queue.empty()) {
+    if (zones_to_clean->empty()) {
```

Comment:
(being a bit paranoiac)  - given that we access zones_to_clean without any locking wouldn't it be more correct to introduce an atomic counter inside ZonedAllocator  to track the amount of entries in this list and permit this thread to proceed? And update it in a single shot in find_zones_to_clean()?
E.g. currently the following scenario is error-prone IMO:
At some iteration this thread calls  
shared_alloc.a->zoned_mark_zones_to_clean_free();

Which (on completion) permits racing ZonedAllocator::allocate() to start filling zones_to_clean with some data.
Which in turn permits this thread to proceed with another zone cleaning iteration (on still incompletely filled zones_to_clean list).
Not sure this might cause any real issues but IMO it's better to avoid such questionable scenarios... Introducing atomic as per above would fix this.

Reply:
Right, I wasn't comfortable with that either. Added an atomic, but I'm not happy with this interface either, the fact that we have two different calls one for accessing the list and another for accessing the list size... Anyway, let me know what you think.

Reply:
Please see another comment inline.